File: HoloLens/DirectX/SampleHoloLens/AnchorExchanger.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
#include "pch.h"
#include "AnchorExchanger.h"
#include "winrt/Windows.Web.Http.h"

using namespace winrt;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::Web::Http;
using namespace winrt::Windows::System::Threading;

AnchorExchanger::AnchorExchanger(const std::wstring& redisUri)
{
    m_redisUri = redisUri;
}


AnchorExchanger::~AnchorExchanger()
{
    if (m_timer != nullptr)
    {
        m_timer.Cancel();
    }
}

std::vector<winrt::hstring> AnchorExchanger::AnchorKeys()
{
    auto lock = std::unique_lock<std::mutex>(m_mutex);
    return m_anchorKeys;
}

winrt::fire_and_forget AnchorExchanger::QueryAndAddAnchorKey(ThreadPoolTimer const&)
{
    auto currentKey = co_await RetrieveLastAnchorKey();
    if (currentKey != L"")
    {
        auto lock = std::unique_lock<std::mutex>(m_mutex);
        if (std::find(m_anchorKeys.begin(), m_anchorKeys.end(), currentKey) == m_anchorKeys.end())
        {
            m_anchorKeys.emplace_back(currentKey);
        }
    }
}

void AnchorExchanger::WatchKeys()
{
    m_timer = ThreadPoolTimer::CreatePeriodicTimer({ this, &AnchorExchanger::QueryAndAddAnchorKey }, std::chrono::milliseconds(500));
}

winrt::Windows::Foundation::IAsyncOperation<hstring> AnchorExchanger::RetrieveAnchorKey(int64_t roomId)
{
    try
    {
        Uri uri{ m_redisUri + L"/" + std::to_wstring(roomId) };
        HttpClient httpClient{};
        return co_await httpClient.GetStringAsync(uri);
    }
    catch (...)
    {
        return {};
    }
}

winrt::Windows::Foundation::IAsyncOperation<hstring> AnchorExchanger::RetrieveLastAnchorKey()
{
    try
    {
        Uri uri{ m_redisUri + L"/last" };
        HttpClient httpClient{};
        return co_await httpClient.GetStringAsync(uri);
    }
    catch (...)
    {
        return {};
    }
}

winrt::Windows::Foundation::IAsyncOperation<int64_t> AnchorExchanger::StoreAnchorKey(const hstring& anchorKey)
{
    co_await winrt::resume_background();
    if (anchorKey.empty())
    {
        return -1;
    }

    {
        auto lock = std::unique_lock<std::mutex>(m_mutex);
        m_anchorKeys.emplace_back(anchorKey);
    }

    try
    {
        HttpClient httpClient{};
        auto response{ co_await httpClient.PostAsync(Uri{ m_redisUri }, HttpStringContent{ anchorKey }) };
        if (response.IsSuccessStatusCode())
        {
            auto responseBody{ co_await response.Content().ReadAsStringAsync() };
            return std::wcstoll(responseBody.c_str(), nullptr, 0);
        }
    }
    catch (...)
    {
        return -1;
    }
}

File: Unity/Assets/AzureSpatialAnchors.Examples/Scripts/AnchorExchanger.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
using System;
using System.Net.Http;
using System.Threading.Tasks;
using UnityEngine;
using System.Collections.Generic;

namespace Microsoft.Azure.SpatialAnchors.Unity.Examples
{
    public class AnchorExchanger
    {
#if !UNITY_EDITOR
        private string baseAddress = "";

        private List<string> anchorkeys = new List<string>();

        public List<string> AnchorKeys
        {
            get
            {
                lock (anchorkeys)
                {
                    return new List<string>(anchorkeys);
                }
            }
        }

        public void WatchKeys(string exchangerUrl)
        {
            baseAddress = exchangerUrl;
            Task.Factory.StartNew(async () =>
                {
                    string previousKey = string.Empty;
                    while (true)
                    {
                        string currentKey = await RetrieveLastAnchorKey();
                        if (!string.IsNullOrWhiteSpace(currentKey) && currentKey != previousKey)
                        {
                            Debug.Log("Found key " + currentKey);
                            lock (anchorkeys)
                            {
                                anchorkeys.Add(currentKey);
                            }
                            previousKey = currentKey;
                        }
                        await Task.Delay(500);
                    }
                }, TaskCreationOptions.LongRunning);
        }

        public async Task<string> RetrieveAnchorKey(long anchorNumber)
        {
            try
            {
                HttpClient client = new HttpClient();
                return await client.GetStringAsync(baseAddress + "/" + anchorNumber.ToString());
            }
            catch (Exception ex)
            {
                Debug.LogException(ex);
                Debug.LogError($"Failed to retrieve anchor key for anchor number: {anchorNumber}.");
                return null;
            }
        }

        public async Task<string> RetrieveLastAnchorKey()
        {
            try
            {
                HttpClient client = new HttpClient();
                return await client.GetStringAsync(baseAddress + "/last");
            }
            catch (Exception ex)
            {
                Debug.LogException(ex);
                Debug.LogError("Failed to retrieve last anchor key.");
                return null;
            }
        }

        internal async Task<long> StoreAnchorKey(string anchorKey)
        {
            if (string.IsNullOrWhiteSpace(anchorKey))
            {
                return -1;
            }

            try
            {
                HttpClient client = new HttpClient();
                var response = await client.PostAsync($"{baseAddress}/key?anchorKey={anchorKey}", new StringContent(string.Empty));
                if (response.IsSuccessStatusCode)
                {
                    string responseBody = await response.Content.ReadAsStringAsync();
                    long ret;
                    if (long.TryParse(responseBody, out ret))
                    {
                        Debug.Log("Key " + ret.ToString());
                        return ret;
                    }
                    else
                    {
                        Debug.LogError($"Failed to store the anchor key. Failed to parse the response body to a long: {responseBody}.");
                    }
                }
                else
                {
                    Debug.LogError($"Failed to store the anchor key: {response.StatusCode} {response.ReasonPhrase}.");
                }

                Debug.LogError($"Failed to store the anchor key: {anchorKey}.");
                return -1;
            }
            catch (Exception ex)
            {
                Debug.LogException(ex);
                Debug.LogError($"Failed to store the anchor key: {anchorKey}.");
                return -1;
            }
        }
#endif
    }
}
